pkg/operators/ebpf/snapshotter.go

```diff
@@ -39,14 +41,53 @@ type Snapshotter struct {
 	ds       datasource.DataSource
 	accessor datasource.FieldAccessor
 	netns    datasource.FieldAccessor
+
+	// iterators is a map of iterators that this snapshotter is composed of. The
+	// key is the iterator name and the value is the linkSnapshotter that
+	// contains the link to the iterator and the iterator type (task, tcp, udp).
+	iterators map[string]*linkSnapshotter
+}
+
+func (i *ebpfInstance) parseSnapshotterPrograms(programs []string) (map[string]*linkSnapshotter, error) {
+	iterators := make(map[string]*linkSnapshotter, len(programs))
+
+programs:
+	for _, program := range programs {
+		if program == "" {
+			return nil, errors.New("empty program name")
+		}
+
+		i.logger.Debugf("> program %q", program)
+
+		// Check if the program is in the eBPF object
+		for _, p := range i.collectionSpec.Programs {
+			if p.Name != program {
+				continue
+			}
+
+			if p.Type != ebpf.Tracing || !strings.HasPrefix(p.SectionName, iterPrefix) {
+				return nil, fmt.Errorf("invalid program %q: expecting type %q and section name prefix %q, got %q and %q",
+					program, ebpf.Tracing, iterPrefix, p.Type, p.SectionName)
+			}
+
+			iterators[program] = nil
```

Comment:
I find this "set it to nil, look it up in another place and fill it" implementation weird. I'm wondering if introducing a new program to snapshotter map in the ebpf instance could help with the overall implementation:

```diff
diff --git a/pkg/operators/ebpf/ebpf.go b/pkg/operators/ebpf/ebpf.go
index 9fe66dd8f..2441510ba 100644
--- a/pkg/operators/ebpf/ebpf.go
+++ b/pkg/operators/ebpf/ebpf.go
@@ -99,10 +99,11 @@ func (o *ebpfOperator) InstantiateImageOperator(
                program: program,

                // Preallocate maps
-               tracers:      make(map[string]*Tracer),
-               structs:      make(map[string]*Struct),
-               snapshotters: make(map[string]*Snapshotter),
-               params:       make(map[string]*param),
+               tracers:               make(map[string]*Tracer),
+               structs:               make(map[string]*Struct),
+               snapshotters:          make(map[string]*Snapshotter),
+               params:                make(map[string]*param),
+               programsToSnashotters: make(map[string]*Snapshotter),

                containers: make(map[string]*containercollection.Container),

@@ -161,6 +162,8 @@ type ebpfInstance struct {

        links []link.Link

+       programsToSnashotters map[string]*Snapshotter
+
        containers map[string]*containercollection.Container

        enums      map[string]*btf.Enum
@@ -547,7 +550,6 @@ func (i *ebpfInstance) Start(gadgetCtx operators.GadgetContext) error {
                }(tracer)
        }

-programs:
        // Attach programs
        for progName, p := range i.collectionSpec.Programs {
                l, err := i.attachProgram(gadgetCtx, p, i.collection.Programs[progName])
@@ -567,19 +569,20 @@ programs:
                                return fmt.Errorf("link is not an iterator")
                        }

-                       for _, snapshotter := range i.snapshotters {
-                               if _, ok := snapshotter.iterators[progName]; ok {
-                                       snapshotter.iterators[progName] = &linkSnapshotter{
-                                               link: lIter,
-                                               typ:  p.AttachTo,
-                                       }
-                                       continue programs
-                               }
+                       snapshotter, ok := i.programsToSnashotters[progName]
+                       if !ok {
+                               // Iterators can be only run by snapshotters
+                               i.Close()
+
+                               // TODO: warning instead?
+                               return fmt.Errorf("iterator %q is not associated with any snapshotter", progName)
                        }

-                       // Iterators can be only run by snapshotters
-                       i.Close()
-                       return fmt.Errorf("iterator %q is not associated with any snapshotter", progName)
+                       snapshotter.iterators = append(snapshotter.iterators, linkSnapshotter{
+                               name: progName,
+                               link: lIter,
+                               typ:  p.AttachTo,
+                       })
                }
        }

diff --git a/pkg/operators/ebpf/snapshotter.go b/pkg/operators/ebpf/snapshotter.go
index cbb46b736..f0787c9c9 100644
--- a/pkg/operators/ebpf/snapshotter.go
+++ b/pkg/operators/ebpf/snapshotter.go
@@ -31,6 +31,7 @@ import (
 )

 type linkSnapshotter struct {
+       name string
        link *link.Iter
        typ  string
 }
@@ -45,39 +46,30 @@ type Snapshotter struct {
        // iterators is a map of iterators that this snapshotter is composed of. The
        // key is the iterator name and the value is the linkSnapshotter that
        // contains the link to the iterator and the iterator type (task, tcp, udp).
-       iterators map[string]*linkSnapshotter
+       iterators []linkSnapshotter
 }

-func (i *ebpfInstance) parseSnapshotterPrograms(programs []string) (map[string]*linkSnapshotter, error) {
-       iterators := make(map[string]*linkSnapshotter, len(programs))
-
-programs:
+func (i *ebpfInstance) parseSnapshotterPrograms(programs []string) error {
        for _, program := range programs {
                if program == "" {
-                       return nil, errors.New("empty program name")
+                       return errors.New("empty program name")
                }

                i.logger.Debugf("> program %q", program)

                // Check if the program is in the eBPF object
-               for _, p := range i.collectionSpec.Programs {
-                       if p.Name != program {
-                               continue
-                       }
-
-                       if p.Type != ebpf.Tracing || !strings.HasPrefix(p.SectionName, iterPrefix) {
-                               return nil, fmt.Errorf("invalid program %q: expecting type %q and section name prefix %q, got %q and %q"
,
-                                       program, ebpf.Tracing, iterPrefix, p.Type, p.SectionName)
-                       }
-
-                       iterators[program] = nil
-                       continue programs
+               p, ok := i.collectionSpec.Programs[program]
+               if !ok {
+                       return fmt.Errorf("program %q not found in eBPF object", program)
                }

-               return nil, fmt.Errorf("program %q not found in eBPF object", program)
+               if p.Type != ebpf.Tracing || !strings.HasPrefix(p.SectionName, iterPrefix) {
+                       return fmt.Errorf("invalid program %q: expecting type %q and section name prefix %q, got %q and %q",
+                               program, ebpf.Tracing, iterPrefix, p.Type, p.SectionName)
+               }
        }

-       return iterators, nil
+       return nil
 }

 func (i *ebpfInstance) populateSnapshotter(t btf.Type, varName string) error {
@@ -105,7 +97,7 @@ func (i *ebpfInstance) populateSnapshotter(t btf.Type, varName string) error {
                i.logger.Debugf("> successfully validated with metadata")
        }

-       iterators, err := i.parseSnapshotterPrograms(parts[2:])
+       err := i.parseSnapshotterPrograms(parts[2:])
        if err != nil {
                return fmt.Errorf("parsing snapshotter %q programs: %w", name, err)
        }
@@ -121,13 +113,18 @@ func (i *ebpfInstance) populateSnapshotter(t btf.Type, varName string) error {
        }

        i.logger.Debugf("adding snapshotter %q", name)
-       i.snapshotters[name] = &Snapshotter{
+       snapshotter := &Snapshotter{
                Snapshotter: metadatav1.Snapshotter{
                        StructName: btfStruct.Name,
                },
-               iterators: iterators,
        }

+       for _, program := range parts[2:] {
+               i.programsToSnashotters[program] = snapshotter
+       }
+
+       i.snapshotters[name] = snapshotter
+
        err = i.populateStructDirect(btfStruct)
        if err != nil {
                return fmt.Errorf("populating struct %q for snapshotter %q: %w", btfStruct.Name, name, err)
@@ -140,19 +137,19 @@ func (i *ebpfInstance) runSnapshotters() error {
        for sName, snapshotter := range i.snapshotters {
                i.logger.Debugf("Running snapshotter %q", sName)

-               for pName, l := range snapshotter.iterators {
-                       i.logger.Debugf("Running iterator %q", pName)
+               for _, l := range snapshotter.iterators {
+                       i.logger.Debugf("Running iterator %q", l.name)
                        switch l.typ {
                        case "task":
                                buf, err := bpfiterns.Read(l.link)
                                if err != nil {
-                                       return fmt.Errorf("reading iterator %q: %w", pName, err)
+                                       return fmt.Errorf("reading iterator %q: %w", l.name, err)
                                }

                                size := snapshotter.accessor.Size()
                                if uint32(len(buf))%size != 0 {
                                        return fmt.Errorf("iter %q returned an invalid buffer's size %d, expected multiple of %d",
-                                               pName, len(buf), size)
+                                               l.name, len(buf), size)
                                }

                                for i := uint32(0); i < uint32(len(buf)); i += size {
@@ -178,13 +175,13 @@ func (i *ebpfInstance) runSnapshotters() error {

                                                buf, err := io.ReadAll(reader)
                                                if err != nil {
-                                                       return fmt.Errorf("reading iterator %q: %w", pName, err)
+                                                       return fmt.Errorf("reading iterator %q: %w", l.name, err)
                                                }

                                                size := snapshotter.accessor.Size()
                                                if uint32(len(buf))%size != 0 {
                                                        return fmt.Errorf("iter %q returned an invalid buffer's size %d, expected multip
le of %d",
-                                                               pName, len(buf), size)
+                                                               l.name, len(buf), size)
                                                }

                                                for i := uint32(0); i < uint32(len(buf)); i += size {
@@ -203,7 +200,7 @@ func (i *ebpfInstance) runSnapshotters() error {
                                        })
                                        if err != nil {
                                                return fmt.Errorf("entering container %q's netns to run iterator %q: %w",
-                                                       container.Runtime.RuntimeName, pName, err)
+                                                       container.Runtime.RuntimeName, l.name, err)
                                        }
                                }
                        }
```

(NOT TESTED)

Reply:
> I find this "set it to nil, look it up in another place and fill it" implementation weird.
>
I used two variables to avoid doing that. But I added both to the snapshot struct.